drivers/regulator/regulator_gpio.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2019-2020 Peter Bigot Consulting, LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT regulator_gpio
+
+#include <drivers/regulator.h>
+#include <drivers/gpio.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(regulator_gpio, CONFIG_REGULATOR_LOG_LEVEL);
+
+#define OPTION_BOOT_ALWAYS_ON_POS 0
+#define OPTION_BOOT_ALWAYS_ON BIT(OPTION_BOOT_ALWAYS_ON_POS)
+#define OPTION_BOOT_ON_POS 0
+#define OPTION_BOOT_ON BIT(OPTION_BOOT_ON_POS)
+
+struct driver_config {
+	const char *gpio_name;
+	uint32_t delay_us;
+	gpio_pin_t gpio_pin;
+	gpio_dt_flags_t gpio_flags;
+	uint8_t options;
+};
+
+struct driver_data {
+	struct device *dev;
+	struct device *gpio;
+	struct onoff_manager mgr;
+};
+
+/* Declaration here, definition after members defined. */
+static const struct regulator_driver_api api;
+
+static void start(struct onoff_manager *mgr,
+		  onoff_notify_fn notify)
+{
+	const struct driver_data *data =
+		CONTAINER_OF(mgr, struct driver_data, mgr);
+	const struct driver_config *cfg = data->dev->config_info;
+	int rc = gpio_pin_set(data->gpio, cfg->gpio_pin, true);
+
+	if (rc >= 0) {
+		k_busy_wait(cfg->delay_us);
```

Comment:
I'm not sure this can happen in ISR.

Reply:
from device runtime pm framework perspective, we can't know in advance whether the power state transition of individual device is ISR friendly or not, so IMHO maybe we can't say the onoff_request can be called in ISR.

Reply:
`k_busy_wait()` is callable from pre-kernel and interrupt contexts, it's just not generally a good idea because it slows things down.

By design and implementation `onoff_request()` can always be called from an ISR.  It's the responsibility of the service's `start` transition function to guarantee that they can be invoked from that context and, if they need to do something that isn't allowed, return OK after putting something on a work queue or handing the operation off to another thread to complete the operation.  Similarly for `onoff_release()` even though that's a synchronous operation.

Here I still need to do that: the `start` and `stop` operations need to check whether `gpio_pin_set()` return `-EWOULDBLOCK`, as they would on an I2C-based IO extender.  That wouldn't affect the regulator API, though.

For the init the code is incomplete: it could check `k_is_pre_kernel()` to determine whether threads are available, and whether the operation completed immediately.  Depending on context the init would either fail, or could proceed and use a semaphore.  (Though It's a little weird to sleep during driver initialization.  I'm not sure the kernel's designed to have multiple threads running (and accessing drivers) at that point, but as long as only the idle thread is available it should work.)

We don't have a plan/API design for turning devices on and off as required by power management or #19448 without power management, so `k_busy_wait()` is fine in the init function ignoring power optimization.

Reply:
> We don't have a plan/API design for turning devices on and off as required by power management

@pabigot I think we need a draft PR first for device runtime PM based on onoff service to do the decision